controllers/installation_controller.go

```diff
@@ -0,0 +1,179 @@
+	var nodes corev1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+	seen := map[string]bool{}
+	needsUpdate := false
+	for _, node := range nodes.Items {
+		seen[node.Name] = true
+		event := metrics.NodeEventFromNode(in.Spec.ClusterID, node)
+		changed, isnew, err := r.NodeHasChanged(in, event)
+		if err != nil {
+			return fmt.Errorf("failed to check if node has changed: %w", err)
+		} else if !changed {
+			continue
+		}
+		needsUpdate = true
+		if err := r.UpdateNodeStatus(in, event); err != nil {
+			return fmt.Errorf("failed to update node status: %w", err)
+		}
+		if isnew {
+			if err := metrics.NotifyNodeAdded(ctx, in.Spec.MetricsBaseURL, event); err != nil {
+				return fmt.Errorf("failed to notify node added: %w", err)
+			}
+			continue
+		}
+		if err := metrics.NotifyNodeUpdated(ctx, in.Spec.MetricsBaseURL, event); err != nil {
+			return fmt.Errorf("failed to notify node updated: %w", err)
+		}
+	}
+	trimmed := []v2alpha1.NodeStatus{}
+	for _, nodeStatus := range in.Status.NodesStatus {
+		if _, ok := seen[nodeStatus.Name]; ok {
+			trimmed = append(trimmed, nodeStatus)
+			continue
+		}
+		needsUpdate = true
+		rmevent := metrics.NodeRemovedEvent{ClusterID: in.Spec.ClusterID, NodeName: nodeStatus.Name}
+		if err := metrics.NotifyNodeRemoved(ctx, in.Spec.MetricsBaseURL, rmevent); err != nil {
+			return fmt.Errorf("failed to notify node removed: %w", err)
+		}
+	}
+	log := ctrl.LoggerFrom(ctx)
+	in.Status.NodesStatus = trimmed
+	if !needsUpdate {
+		log.Info("No node changes detected")
+		return nil
+	}
+	log.Info("Node changes detected, updating installation status")
+	if err := r.Status().Update(ctx, in); err != nil {
+		return fmt.Errorf("failed to update installation status: %w", err)
+	}
+	return nil
+}
+
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+//+kubebuilder:rbac:groups=cluster.kurl.sh,resources=installations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.kurl.sh,resources=installations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cluster.kurl.sh,resources=installations/finalizers,verbs=update
+
+// Reconcile reconcile the installation object.
+func (r *InstallationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	var installs v2alpha1.InstallationList
+	if err := r.List(ctx, &installs); err != nil {
+		return ctrl.Result{}, err
+	}
+	log.Info("Reconciling installation")
+	if len(installs.Items) == 0 {
+		log.Info("No installation found, reconciliation ended")
+		return ctrl.Result{}, nil
+	}
+	if len(installs.Items) > 1 {
+		log.Error(fmt.Errorf("multiple installations found"), "Multiple installations found")
+		return ctrl.Result{RequeueAfter: requeueAfter}, nil
+	}
+	if installs.Items[0].Spec.AirGap {
+		log.Info("Airgap installation, reconciliation ended")
+		return ctrl.Result{}, nil
+	}
```

Comment:
should we still reconcile but not send metrics on airgap?

Reply:
Yes. Fixed.